mantidproject/mantid
# Remove ADS from MDGeometry and follow-up include cleanup

This PR does a series of cleanup for `#include`:
- `AnalysisDataService.h` removed from from `MDGeometry.h` (which is a base class of `MatrixWorkspace`, so it is included a *lot*).
- `WorkspaceHistory.h` include is removed from `Workspace.h`.
- Moved definition of `VMDBase` into source to reduce number of required includes.
- Several "trivial" moves of includes, possible by using forward declarations.

This is a follow up to https://github.com/mantidproject/mantid/pull/17860. Wait with merging until that PR is merged.

This PR brings in the harvest after the preparations done in https://github.com/mantidproject/mantid/pull/17860 (which was more effort):

These changes reduce the average number of distinct recursively included files for cpp files in `Framework/` from 107 to 90 (code outside `Framework/` is also affected, but I did not collect statistics there).

As a more detailed example, result for `ConvertUnits.cpp` are (GCC 5.4, Linux Workstation):

Before PR17860:
```
Includes: 198
Compile-time: 7.0 seconds
Object size: 691256 Byte
```
After this PR:
```
Includes: 159
Compile-time: 5.4 seconds
Object size: 484032 Byte
```
That is, roughly 20% improvement in build time.


**To test:**

Note that github shows also the the changes from PR17860. For reviewing make sure to look only at the changes since then (in the files tab, select commits on the top left).

Significant code changes are in `MDGeometry`, `VMD`, `ParameterMap`, `LogManager`, `Workspace`, and `Run`, please review those carefully. Apart from that it is mostly includes that change.
Fixes #17918.
Internal change, no release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
